src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,505 @@
+        modeling). Can be set to None if all positions should be used. `peft.tuners.lora.eva.get_indices_fn_causal_lm`
+        is used by default. Should always return a tensor of shape (n_indices, layer_input.ndim-1).
+    prepare_activations_fn: (Union[callable, Dict[str, callable], None])
+        If a layer receives multiple inputs as a list, per default the first input is used. This function can be used
+        to modify this behaviour. Accepts a dictionary with layer names as keys. Default logic:
+        ```
+        try:
+            svd_input = activations.detach()
+        except AttributeError:
+            svd_input = activations[0].detach()
+        ```
+
+    Returns:
+    -------- eva_state_dict: (dict)
+        The state dictionary containing the SVD components for each layer.
+    """
+
+    # Computes the rank distribution for each layer based on the explained variance ratio.
+    def _get_rank_distribution(hooks, layer_hook_map, equal_inputs_map, rank_budget, max_components):
+        exp_vars = {k: h[0].svd.explained_variance_ratio_[:max_components] for k, h in hooks.items()}
+        keys, values = zip(*[(k, c) for k, name in layer_hook_map.items() for c in exp_vars[name]])
+        idx = torch.stack(values).argsort(descending=True)
+        counts = Counter([keys[i] for i in idx[:rank_budget]])
+        counts = {k: counts.get(k, 0) for k in layer_hook_map.keys()}  # add layers with 0 rank
+        for k, k_hook in equal_inputs_map.items():
+            # ensure hook layers have the highest rank if they are equal to another layer
+            rank, rank_hook = counts[k], counts[k_hook]
+            if rank_hook >= rank:
+                continue
+            counts[k_hook], counts[k] = rank, rank_hook
+        return counts
+
+    # set function to check if modules should be added to hooks
+    if hasattr(model, "peft_config"):
+
+        def _check_fn(name, module):
+            return isinstance(module, Linear)
+    else:
+
+        def _check_fn(name, module):
+            return any(name.endswith(t) for t in peft_config.target_modules)
+
+    training = model.training
+    device = get_device_with_meta_params(model)
+    model.eval()
+
+    hooks = {}
+    for name, module in model.named_modules():
+        # currently only linear layers are supported
+        if not _check_fn(name, module):
+            continue
+        hook = HashHook(name)
+        handle = module.register_forward_hook(hook)
+        hooks[name] = (hook, handle)
+    rank_budget = peft_config.r * len(hooks)
+    max_components = round(peft_config.r * peft_config.eva_config.rho)
+
+    # forward for one batch to check which layer inputs are equal to avoid unneeded svd calculations
+    inputs = next(iter(dataloader))
+    forward_fn(model, move_inputs_to_device(inputs, device))
+    hash_dict = {k: h[0].hashed_inputs[0] for k, h in hooks.items()}
+    equal_inputs_map = {vv: v[0] for v in find_equal_values(hash_dict).values() for vv in v[1:]}
```

Comment:
I don't really understand what should happen to have `equal_inputs_map`, could you please add a comment?

Reply:
Added a comment. It contains information about which layers recieve the same input activations (e.g. queries, keys and values). For those groups we only need to compute the svd once.